Forta-Bots-Challenge-2/src/agent.spec.ts

```diff
@@ -0,0 +1,216 @@
+    mockProvider.addCallTo(token0, block, ERC20_IFACE, "symbol", {
+      inputs: [],
+      outputs: ["TKN0"],
+    });
+    mockProvider.addCallTo(token1, block, ERC20_IFACE, "symbol", {
+      inputs: [],
+      outputs: ["TKN1"],
+    });
+  };
+
+  beforeAll(() => {
+    handleTransaction = provideHandleTransaction(SWAP_EVENT);
+  });
+
+  beforeEach(async () => {
+    initialize = provideInitialize(mockProvider as any, mockUniswapFactory, mockInitCodeHash);
+    await initialize();
+    mockTxEvent = new TestTransactionEvent();
+  });
+
+  it("returns empty findings if there is an empty txEvent", async () => {
+    const findings = await handleTransaction(mockTxEvent);
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("returns empty findings if there is a swap on a pool with same tokens and fee deployed by a different factory", async () => {
+    const wrongFactoryAddress = createAddress("0x03");
+    const [block, token0, token1, fee] = TEST_DATA[0];
+
+    const badPoolAddress = getPoolAddress(wrongFactoryAddress, mockInitCodeHash, [token0, token1, fee]);
+
+    mockPoolCalls(badPoolAddress, block, UNISWAP_POOL_IFACE, [
+      ["token0", token0],
+      ["token1", token1],
+      ["fee", fee],
+    ]);
+
+    mockERC20Calls(token0, token1, block);
+
+    mockTxEvent.setBlock(block).addEventLog(SWAP_EVENT, badPoolAddress, TEST_LOGS[0]);
+
+    const findings = await handleTransaction(mockTxEvent);
+
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("returns empty findings if there is a swap on a pool with incorrect ABI", async () => {
+    const wrongInitCodeHash = keccak256(createAddress("0x03"));
+    const [block, token0, token1, fee] = TEST_DATA[0];
+
+    const poolAddress = getPoolAddress(mockUniswapFactory, wrongInitCodeHash, [token0, token1, fee]);
+
+    mockPoolCalls(poolAddress, block, INCORRECT_IFACE, [["foo", token0]]);
+
+    mockERC20Calls(token0, token1, block);
+
+    mockTxEvent.setBlock(block).addEventLog(SWAP_EVENT, poolAddress, TEST_LOGS[0]);
+
+    const findings = await handleTransaction(mockTxEvent);
+
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("returns correct findings if there is one swap on Uniswap V3", async () => {
+    const [block, token0, token1, fee] = TEST_DATA[0];
+
+    const poolAddress = getPoolAddress(mockUniswapFactory, mockInitCodeHash, [token0, token1, fee]);
+
+    mockPoolCalls(poolAddress, block, UNISWAP_POOL_IFACE, [
+      ["token0", token0],
+      ["token1", token1],
+      ["fee", fee],
+    ]);
+
+    mockERC20Calls(token0, token1, block);
+
+    mockTxEvent.setBlock(block).addEventLog(SWAP_EVENT, poolAddress, TEST_LOGS[0]);
+
+    const findings = await handleTransaction(mockTxEvent);
+
+    expect(findings).toStrictEqual([
+      expect.objectContaining(createTestFinding(TEST_LOGS[0], poolAddress, TEST_POOL_DATA_0)),
+    ]);
+  });
+
+  it("returns correct findings if there are multiple swaps from multiple pools on Uniswap V3", async () => {
```

Comment:
Would recommend adding one more test case where there are multiple swaps, both from correct contracts and incorrect contracts, but the bot **only** detects the correct ones. Just to confirm it could handle both types in the same payload.